WI #616: Addition of five columns related to `Pricing Currency` (#848)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/columns/pricingcurrency.md

# Pricing Currency

[*Pricing Currency*](#glossary:pricing-currency) is the national or virtual currency denomination that a [*resource*](#glossary:resource) or [*service*](#glossary:service) was priced in. Pricing Currency is commonly used in scenarios where different currencies are used for pricing and billing.

The PricingCurrency column adheres to the following requirements:

* PricingCurrency MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports pricing and billing in different currencies.
* PricingCurrency MUST be of type String.
* PricingCurrency MUST conform to [StringHandling](#stringhandling) requirements.
* PricingCurrency MUST conform to [CurrencyFormat](#currencyformat) requirements.
* PricingCurrency MUST NOT be null.

## Column ID

PricingCurrency

## Display Name

Pricing Currency

## Description

> Represents the unit of measure in which a charge was priced.

Reviewer: ```suggestion
The national or virtual currency denomination that a *resource* or *service* was priced in.
```

# Pricing Currency

[*Pricing Currency*](#glossary:pricing-currency) is a value that represents the denomination in which a charge for [*resources*](#glossary:resource) or [*services*](#glossary:service) was priced. Pricing Currency is commonly used in scenarios where costs need to be grouped or aggregated by some combination of the following currency types:

* National currency (e.g. USD, EUR).
* Virtual currency (e.g. tokens, credits).

The PricingCurrency column adheres to the following requirements:

* PricingCurrency presence in a [*FOCUS dataset*](#glossary:FOCUS-dataset) is defined as follows:

>   * PricingCurrency MUST be present in a FOCUS dataset when the provider presents prices in a virtual currency (e.g. credits, tokens).

Reviewer 2: I believe `Pricing{Cost,Currency}` should be designated in the provider's list pricing currency unit (e.g. credits, tokens). If the provider additionally provides a list cost per pricing currency unit (e.g. $1 per token) but the `Billing{Cost,Currency}` is in another national currency (e.g. EUR), I imagine both the purchase of tokens and usage records would designate EUR as the `Billing{Cost,Currency}`. Given both the purchase and usage records must be designated in the same `Billing{Cost,Currency}`, I believe the exchange rate column could account for both exchange rate translations (i.e. token -> USD -> EUR). 

Moreover, the theoretical intermediate national currency list pricing exchange rate (i.e. $1 per token) seems an implementation detail as providers could choose to provide list exchange rates from their virtual currency to any supported national currencies as needed (e.g. €0.92 per token). 

Also, given these exchange rates may additionally change between the time of purchase and usage I think we should fully expect these exchange rates to differ across rows even for the same translation (e.g. token -> USD).

I don't think the spec needs to explain all of these translations at various points in time, it just needs to provide an exchange between the `Pricing{Cost,Currency}` and `Billing{Cost,Currency}` which can (a) vary overtime and (b) account for multiple currency translations.

>   * PricingCurrency MUST be present in a FOCUS dataset when the provider presents prices in a virtual currency (e.g. credits, tokens).

Reviewer 3: I agree with @toddkblake that all sounds reasonable. I really like the fact this set of columns has more value than just virtual currency and actually solves the multi-national currency issue for existing FOCUS datasets.

>   * PricingCurrency MUST be present in a FOCUS dataset when the provider presents prices in a virtual currency (e.g. credits, tokens).

Reviewer 4: I'll look at the examples that @mike-finopsorg has provided and follow up (quickly) on this thread with either questions or resolving it.

>   * PricingCurrency MUST be present in a FOCUS dataset when the provider presents prices in a virtual currency (e.g. credits, tokens).

Author: Resolving per discussions in TF3 on Mar 21, and per confirmation from @cnharris10 in Maintainers on Mar 24.

>   * PricingCurrency MAY be present in a FOCUS dataset in all other cases.

Author: @ijurica Did I need to apply your simplified presence suggestion here as well?

Reviewer 5: Yes, it seems to me that the same condition applies to all of these columns.

---

I wasn't able to address all the bullets in a single suggestion because there was a Git comment in between, so I just left a note [here](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/848#discussion_r2005369101):

> Since the group of normative requirements is split by the conversation, if we decide to go with this version, the remaining two presence-related requirements should also be removed.

Author: ```suggestion
* PricingCurrency MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports pricing and billing in different currencies.
```

>   * PricingCurrency MUST be present in a FOCUS dataset when the provider presents prices and bills in different fiat currencies (e.g. priced in USD and billed in EUR).

Reviewer 6: ```suggestion
  * PricingCurrency MUST be present in a FOCUS dataset when the provider presents prices and usage in different national currencies (e.g. priced in USD and billed in EUR).
```

Author: Great catch on `fiat` vs `national`.  Sorry I missed it in my overall scrub.

However, when I used the word `bills`, it was in the verb form, not noun.  

So with both of those considerations in mind, does this revision work for you, @thecloudman?

```suggestion
  * PricingCurrency MUST be present in a FOCUS dataset when the provider bills in different national currencies (e.g. priced in USD and billed in EUR).
```

Author: Actually, upon second look, you're right, but then I think we need to still make a small change.  How about:
```suggestion
  * PricingCurrency MUST be present in a FOCUS dataset when the provider presents prices and usage in different national currencies (e.g. prices in USD and usage in EUR).
```

  * PricingCurrency MAY be present in a FOCUS dataset in all other cases.

> * PricingCurrency MUST be of type String.

Reviewer 5: nit:

```suggestion
* PricingCurrency MUST be of type String.
* PricingCurrency MUST conform to [StringHandling](#stringhandling) requirements.
```

Author: Yes, thank you!

> * PricingCurrency MUST conform to [Currency Format](#currencyformat) requirements.

Reviewer 5: editorial nit:

```suggestion
* PricingCurrency MUST conform to [CurrencyFormat](#currencyformat) requirements.
```

# Pricing Currency

> [*Pricing Currency*](#glossary:pricing-currency) is a value that represents the denomination in which a charge for [*resources*](#glossary:resource) or [*services*](#glossary:service) was priced. Pricing Currency is commonly used in scenarios where costs need to be grouped or aggregated by some combination of the following currency types:
> 
> * National currency (e.g. USD, EUR).
> * Virtual currency (e.g. tokens, credits).
> 

Reviewer: ```suggestion
[*Pricing Currency*](#glossary:pricing-currency) is the national or virtual currency denomination that a [*resource*](#glossary:resource) or [*service*](#glossary:service) was priced in. Pricing Currency is commonly used in scenarios where different currencies are used for pricing and billing.

```

Author: Same comment as on Pricing To Billing Exchange Rate [here](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/848/files/e45f1d1f32affa42ce89a0ab3c676aeeec505c94#r2004779197).

> Interesting, @UdamDewaraja. Do you think it's confusing or duplicative to be explicit about the kinds of currencies we're facilitating exchange between? I can get on board with striking that language, but I was trying to give context. Curious what others think.

Reviewer: Responding on that one.

Author: I don't yet see your response there; could you link me in case I'm being dense?

Author: Discussed in Mar 21 TF3 and agreed to sustain!

The PricingCurrency column adheres to the following requirements:

> * PricingCurrency presence in a [*FOCUS dataset*](#glossary:FOCUS-dataset) is defined as follows:
>   * PricingCurrency MUST be present in a FOCUS dataset when the provider presents prices in a virtual currency (e.g. credits, tokens).

Reviewer 5: As mentioned [here](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/848#issuecomment-2736862353), the original version slightly implies that providers present all prices (including List and Contracted Unit Price) in the FOCUS dataset in a currency different from the billing currency.

Note: The same pattern applicable to all columns added in this PR.

```suggestion
* PricingCurrency MUST be present in a FOCUS dataset when the provider supports pricing and billing in different currencies.
```

**Note:** 
* Since the group of normative requirements is split by a conversation, if we decide to go with this version, the remaining two presence-related requirements should also be removed.
